experimental/casemapping/src/internals.rs

```diff
@@ -0,0 +1,1066 @@
+//       12..13  Dot type
+//           14  Has conditional special casing
+//           15  Has conditional case folding
+//
+//   If double-width slots is false:
+//
+//   [idx + 1] First optional slot
+//   [idx + 2] Second optional slot
+//   [idx + 3] Third optional slot
+//   ...
+//
+//   If double-width slots is true:
+//
+//   [idx + 1] First optional slot
+//   [idx + 3] Second optional slot
+//   [idx + 5] Third optional slot
+//   ...
+//
+//   Full mappings: If there is at least one full (string) case mapping, then the lengths
+//   of the mappings are encoded as nibbles in the full mappings slot:
+//       Bits:
+//          0..4   Length of lowercase string
+//          5..7   Length of case mapping string
+//          8..11  Length of uppercase string
+//          12..15 Length of titlecase string
+//   Mappings that do not exist have length 0. The strings themselves are stored in the
+//   above order immediately following the last optional slot, encoded as UTF16.
+//
+//   Case closure mappings: If the case closure for a code point includes code points
+//   that are not included in the simple or full mappings, then bits 0..3 of the closure
+//   mappings slot will contain the number of codepoints in the closure string. (Other
+//   bits are reserved.) The closure string itself is encoded as UTF16 and stored
+//   following the full mappings data (if it exists) or the final optional slot.
+//
+//   This representation is generated by casepropsbuilder.cpp in ICU.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Debug, Eq, PartialEq, Clone, Yokeable, ZeroCopyFrom)]
+struct CaseMappingExceptions<'data> {
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    raw: ZeroVec<'data, u16>,
+}
+
+#[derive(Copy, Clone, Debug)]
+enum CaseMappingExceptionSlot {
+    Lower = 0,
+    Fold = 1,
+    Upper = 2,
+    Title = 3,
+    Delta = 4,
+    // Slot 5 is reserved
+    Closure = 6,
+    FullMappings = 7,
+}
+
+impl CaseMappingExceptionSlot {
+    fn contains_char(&self) -> bool {
+	matches!(self, Self::Lower | Self::Fold | Self::Upper | Self::Title)
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+enum FullMapping {
+    Lower = 0,
+    Fold = 1,
+    Upper = 2,
+    Title = 3,
+}
+
+impl<'data> CaseMappingExceptions<'data> {
+    const SLOTS_MASK: u16 = 0xff;
+
+    // Each slot is 2 u16 elements instead of 1
+    const DOUBLE_SLOTS_FLAG: u16 = 0x100;
+
+    const NO_SIMPLE_CASE_FOLDING_FLAG: u16 = 0x200;
+    const DELTA_IS_NEGATIVE_FLAG: u16 = 0x400;
+    const SENSITIVE_FLAG: u16 = 0x800;
+
+    const DOT_SHIFT: u16 = 12;
+
+    const CONDITIONAL_SPECIAL_FLAG: u16 = 0x4000;
+    const CONDITIONAL_FOLD_FLAG: u16 = 0x8000;
+
+    const FULL_MAPPINGS_ALL_LENGTHS_MASK: u32 = 0xffff;
+    const FULL_MAPPINGS_LENGTH_MASK: u32 = 0xf;
+    const FULL_MAPPINGS_LENGTH_SHIFT: u32 = 4;
+
+    const CLOSURE_MAX_LENGTH: u32 = 0xf;
+
+    fn try_from_icu(raw: &[u16]) -> Result<(Self, Vec<u16>), Error> {
```

Comment:
Nit: Hide transformer-specific code behind the feature `provider_transform_internals`